docs/machine-learning/how-to-guides/explain-machine-learning-model-permutation-feature-importance-ml-net.md

```diff
@@ -0,0 +1,159 @@
+string[] featureColumnNames = 
+    data.Schema
+        .Select(column => column.Name)
+        .Where(columnName => columnName != "Label").ToArray();
+
+// 3. Define estimator with data pre-processing steps
+IEstimator<ITransformer> dataPrepEstimator =
+    mlContext.Transforms.Concatenate("Features", featureColumnNames)
+        .Append(mlContext.Transforms.NormalizeMinMax("Features"));
+
+// 4. Create transformer using the data pre-processing estimator
+ITransformer dataPrepTransformer = dataPrepEstimator.Fit(trainData);
+
+// 5. Pre-process the training and test sets
+IDataView preprocessedTrainData = dataPrepTransformer.Transform(trainData);
+IDataView preprocessedTestData = dataPrepTransformer.Transform(testData);
+
+// 6. Define Stochastic Dual Coordinate Ascent machine learning estimator
+var sdcaEstimator = mlContext.Regression.Trainers.Sdca();
+
+// 7. Train machine learning model
+var sdcaModel = sdcaEstimator.Fit(preprocessedTrainData);
+```
+
+## Explain the model with Permutation Feature Importance (PFI)
+
+Permutation Feature Importance (PFI) is a technique used to explain classification and regression models that is inspired by [Breiman's *Random Forests* paper](https://www.stat.berkeley.edu/~breiman/randomforest2001.pdf)(see section 10). At a high level, the way it works is by randomly shuffling data one feature at a time and calculating how much the performance metric of interest decreases. The larger the change, the more important that feature is.
+
+In ML.NET use the [`PermutationFeatureImportance`](xref:Microsoft.ML.PermutationFeatureImportanceExtensions) method for your respective task.
+
+```csharp
+ImmutableArray<RegressionMetricsStatistics> permutationFeatureImportance = 
+    mlContext
+        .Regression
+        .PermutationFeatureImportance(sdcaModel, preprocessedTestData, permutationCount:3);
+```
+
+The result of the operation will be an [`ImmutableArray`](xref:System.Collections.Immutable.ImmutableArray) of [`RegressionMetricsStatistics`](xref:Microsoft.ML.Data.RegressionMetricsStatistics) type. [`RegressionMetricsStatistics`](xref:Microsoft.ML.Data.RegressionMetricsStatistics) provides summary statistics like mean and standard deviation for multiple observations of [`RegressionMetrics`](xref:Microsoft.ML.Data.RegressionMetrics) equal to the number of permutations specified in the `permutationCount` parameter.
+
+For interpretability, the results of PFI can then be ordered based on importance.
+
+```csharp
+// Order features by importance
+var featureImportanceMetrics =
+    permutationFeatureImportance
+        .Select((metric, index) => new { index, metric.RSquared })
+        .OrderByDescending(myFeatures => myFeatures.RSquared.Mean);
+
+Console.WriteLine("Feature\tPFI");
+foreach (var feature in featureImportanceMetrics)
+{
+    Console.WriteLine($"{featureColumnNames[feature.index],-20}|\t{feature.RSquared.Mean,10}");
+}
+```
+
+The output would look similar to that below:
```

Comment:
Did we use a standard data set here? Should we mention that, so that it is easier to understand what the results are?

Reply:
Yes and no. Some of the features are similar to those in the Boston Housing Dataset but the data itself is all generated.

Reply:
#resolved